test/e2e/framework/framework.go

```diff
@@ -200,26 +202,28 @@ func (f *Framework) setupNamespace(ctx context.Context) {
 	userSA, err = WaitForServiceAccount(ctxUserSa, f.KubeAdminClient().CoreV1(), userSA.Namespace, userSA.Name)
 	o.Expect(err).NotTo(o.HaveOccurred())
 
-	// Create a restricted client using the default SA.
-	var token []byte
-	for _, secretName := range userSA.Secrets {
-		secret, err := f.KubeAdminClient().CoreV1().Secrets(userSA.Namespace).Get(ctx, secretName.Name, metav1.GetOptions{})
-		o.Expect(err).NotTo(o.HaveOccurred())
+	// Create a service account token Secret for the user ServiceAccount.
+	userSATokenSecret, err := f.KubeAdminClient().CoreV1().Secrets(ns.Name).Create(ctx, &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: ServiceAccountTokenSecretName,
+			Annotations: map[string]string{
+				corev1.ServiceAccountNameKey: userSA.Name,
+			},
+		},
+		Type: corev1.SecretTypeServiceAccountToken,
+	}, metav1.CreateOptions{})
+	o.Expect(err).NotTo(o.HaveOccurred())
 
-		if secret.Type == corev1.SecretTypeServiceAccountToken {
-			name := secret.Annotations[corev1.ServiceAccountNameKey]
-			uid := secret.Annotations[corev1.ServiceAccountUIDKey]
-			if name == userSA.Name && uid == string(userSA.UID) {
-				t, found := secret.Data[corev1.ServiceAccountTokenKey]
-				if found {
-					token = t
-					break
-				}
-			}
-		}
-	}
-	o.Expect(token).NotTo(o.HaveLen(0))
+	By(fmt.Sprintf("Waiting for service account token Secret %q in namespace %q.", userSATokenSecret.Name, userSATokenSecret.Namespace))
+	ctxUserSATokenSecret, ctxUserSATokenSecretCancel := watchtools.ContextWithOptionalTimeout(ctx, serviceAccountTokenSecretWaitTimeout)
+	defer ctxUserSATokenSecretCancel()
+	userSATokenSecret, err = WaitForServiceAccountTokenSecret(ctxUserSATokenSecret, f.KubeAdminClient().CoreV1(), userSATokenSecret.Namespace, userSATokenSecret.Name)
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	token := userSATokenSecret.Data[corev1.ServiceAccountTokenKey]
```

Comment:
nit: you should check the map has the key

Reply:
Looking up a non-existent key will return a zero value, in this case an empty array, and I'm asserting non-emptiness in the next line, so would that really make any difference? Unless you just mean it for the sake of explicitness?

Reply:
If the next line fails and you are debugging an e2e test there are 2 possible ways how it could have failed. It either didn't have the key or the value is really empty. And you start wondering which it is while you could have just known if the previous line used `o.HaveKey()`